File: backend/routes/purchase.py
"""Purchase order routes — CRUD + confirm + receive + cancel."""
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from database.session import get_db
from dependencies.auth import get_current_user, require
from dependencies.permissions import P
from models.user import User
from schemas.purchase import (
    PurchaseOrderCreate, PurchaseOrderOut, ReceiptRequest,
)
from services.purchase_service import PurchaseService
from utils.enums import PurchaseOrderStatus

router = APIRouter(prefix="/purchase-orders", tags=["Purchase Orders"])


@router.get("", response_model=list[PurchaseOrderOut],
            dependencies=[Depends(require(P.PURCHASE_VIEW))])
def list_purchase(status_filter: PurchaseOrderStatus | None = None,
                  db: Session = Depends(get_db)):      
    return PurchaseService(db).list(status=status_filter)


@router.post("", response_model=PurchaseOrderOut, status_code=status.HTTP_201_CREATED,
             dependencies=[Depends(require(P.PURCHASE_CREATE))])
def create_purchase(payload: PurchaseOrderCreate, db: Session = Depends(get_db),
                    current_user: User = Depends(get_current_user)):
    return PurchaseService(db).create(payload, user_id=current_user.id)


@router.get("/{po_id}", response_model=PurchaseOrderOut,
            dependencies=[Depends(require(P.PURCHASE_VIEW))])
def get_purchase(po_id: int, db: Session = Depends(get_db)):
    return PurchaseService(db).get(po_id)


@router.post("/{po_id}/confirm", response_model=PurchaseOrderOut,
             dependencies=[Depends(require(P.PURCHASE_CONFIRM))])
def confirm_purchase(po_id: int, db: Session = Depends(get_db),
                     current_user: User = Depends(get_current_user)):
    return PurchaseService(db).confirm(po_id, user_id=current_user.id)


@router.post("/{po_id}/receive", response_model=PurchaseOrderOut,
             dependencies=[Depends(require(P.PURCHASE_EDIT))])
def receive_purchase(po_id: int, payload: ReceiptRequest = ReceiptRequest(),
                     db: Session = Depends(get_db),
                     current_user: User = Depends(get_current_user)):
    """Goods receipt: increases on_hand (supports partial receipts)."""
    return PurchaseService(db).receive(po_id, payload, user_id=current_user.id)


@router.post("/{po_id}/cancel", response_model=PurchaseOrderOut,
             dependencies=[Depends(require(P.PURCHASE_DELETE))])
def cancel_purchase(po_id: int, db: Session = Depends(get_db),
                    current_user: User = Depends(get_current_user)):
    return PurchaseService(db).cancel(po_id, user_id=current_user.id)

File: frontend/src/api/purchase.ts
import type {
  PurchaseOrder,
  PurchaseOrderCreate,
  PurchaseOrderStatus,
  ReceiptRequest,
} from "@/types/purchase";

import { api } from "./axios";

export const listPurchaseOrders = (status?: PurchaseOrderStatus): Promise<PurchaseOrder[]> =>
  api
    .get<PurchaseOrder[]>("/purchase-orders", { params: status ? { status_filter: status } : undefined })
    .then((r) => r.data);

export const getPurchaseOrder = (id: number): Promise<PurchaseOrder> =>
  api.get<PurchaseOrder>(`/purchase-orders/${id}`).then((r) => r.data);

export const createPurchaseOrder = (body: PurchaseOrderCreate): Promise<PurchaseOrder> =>
  api.post<PurchaseOrder>("/purchase-orders", body).then((r) => r.data);

export const confirmPurchaseOrder = (id: number): Promise<PurchaseOrder> =>
  api.post<PurchaseOrder>(`/purchase-orders/${id}/confirm`).then((r) => r.data);

export const receivePurchaseOrder = (id: number, body: ReceiptRequest): Promise<PurchaseOrder> =>
  api.post<PurchaseOrder>(`/purchase-orders/${id}/receive`, body).then((r) => r.data);

export const cancelPurchaseOrder = (id: number): Promise<PurchaseOrder> =>
  api.post<PurchaseOrder>(`/purchase-orders/${id}/cancel`).then((r) => r.data);
